<a href="https://colab.research.google.com/github/Ferperezm28/Calidad_Sue-o_DA/blob/main/Calidad_sueno.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Analisis de Calidad de sueño**


# Preguntas a responder:

* ¿Quiénes tienen peor calidad del sueño, hombres o mujeres?
* ¿Existe una relación entre la calidad del sueño de las personas y su profesión?
* ¿La actividad física afecta la calidad de sueño?
* ¿Qué profesión presenta la peor calidad de sueño?
* ¿Cuál profesión tienen a las personas con el mayor nivel de estrés y el mayor índice de masa corporal?
* ¿En qué rango de edades se encuentran la mayor cantidad de trastornos del sueño?
* La cantidad de pasos al día, ¿afecta la calidad del sueño? ¿al índice de masa corporal?
* Las personas que realizan mas actividad fisica, presentan mejor calidad y tiempo de sueño?
* La presion arterial y la frecuencia cardiaca determina la calidad de sueño?


Con estas preguntas comenzaremos a hacer nuestro análisis buscando extraer la mayor cantidad de información que se pueda del conjunto,


Dataset - Sleep Health and Lifestyle Dataset

https://www.kaggle.com/datasets/uom190346a/sleep-health-and-lifestyle-dataset



**Variables en el dataset**

* ID- identificador unico por usuario
* Age - edad del participante
* Occupation - Profesion
* Sleep Duration - Horas que durmio
* Quality of sleep - Calidad de sueño (1-10)
* Physical Activity Level - minutos al dia que la persona estuvo activa
* Stress Level - nivel de estres(1-10)
* BMI Category - Indice de masa corporal
* Blood Preasure - Presion arterial
* Heart rate - Frecuencia cardiaca
* Daily Steps - Pasos diarios
* Sleep Disorder - Desorden del sueño (no, insomnio, Apnea del sueño)

In [1]:
#importar librerias
import pandas as pd

In [2]:
url_repos = 'https://raw.githubusercontent.com/Ferperezm28/Calidad_Sue-o_DA/main/sleep_health.csv'

In [3]:
#cargar dataset
data = pd.read_csv(url_repos, index_col='Person ID')
#asignar ID como indice del df
data.head()

Gender  Age            Occupation  Sleep Duration  Quality of Sleep  \
Person ID                                                                       
1           Male   27     Software Engineer             6.1                 6   
2           Male   28                Doctor             6.2                 6   
3           Male   28                Doctor             6.2                 6   
4           Male   28  Sales Representative             5.9                 4   
5           Male   28  Sales Representative             5.9                 4   

           Physical Activity Level  Stress Level BMI Category Blood Pressure  \
Person ID                                                                      
1                               42             6   Overweight         126/83   
2                               60             8       Normal         125/80   
3                               60             8       Normal         125/80   
4                               30             8        Obese         140/90   
5                               30             8        Obese         140/90   

           Heart Rate  Daily Steps Sleep Disorder  
Person ID                                          
1                  77         4200           None  
2                  75        10000           None  
3                  75        10000           None  
4                  85         3000    Sleep Apnea  
5                  85         3000    Sleep Apnea

In [4]:
df = data.copy()

*EDA*


In [12]:
df.shape

(374, 12)

In [16]:
df.columns

Index(['Gender', 'Age', 'Occupation', 'Sleep Duration', 'Quality of Sleep',
       'Physical Activity Level', 'Stress Level', 'BMI Category',
       'Blood Pressure', 'Heart Rate', 'Daily Steps', 'Sleep Disorder'],
      dtype='object')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 374 entries, 1 to 374
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Gender                   374 non-null    object 
 1   Age                      374 non-null    int64  
 2   Occupation               374 non-null    object 
 3   Sleep Duration           374 non-null    float64
 4   Quality of Sleep         374 non-null    int64  
 5   Physical Activity Level  374 non-null    int64  
 6   Stress Level             374 non-null    int64  
 7   BMI Category             374 non-null    object 
 8   Blood Pressure           374 non-null    object 
 9   Heart Rate               374 non-null    int64  
 10  Daily Steps              374 non-null    int64  
 11  Sleep Disorder           374 non-null    object 
dtypes: float64(1), int64(6), object(5)
memory usage: 38.0+ KB


Caracteristicas de las columnas:

In [13]:
#columnas numericas
df.describe()

Age  Sleep Duration  Quality of Sleep  Physical Activity Level  \
count  374.000000      374.000000        374.000000               374.000000   
mean    42.184492        7.132086          7.312834                59.171123   
std      8.673133        0.795657          1.196956                20.830804   
min     27.000000        5.800000          4.000000                30.000000   
25%     35.250000        6.400000          6.000000                45.000000   
50%     43.000000        7.200000          7.000000                60.000000   
75%     50.000000        7.800000          8.000000                75.000000   
max     59.000000        8.500000          9.000000                90.000000   

       Stress Level  Heart Rate   Daily Steps  
count    374.000000  374.000000    374.000000  
mean       5.385027   70.165775   6816.844920  
std        1.774526    4.135676   1617.915679  
min        3.000000   65.000000   3000.000000  
25%        4.000000   68.000000   5600.000000  
50%        5.000000   70.000000   7000.000000  
75%        7.000000   72.000000   8000.000000  
max        8.000000   86.000000  10000.000000

Estadisticas de las columnas numericas:

1. *AGE*: El promedio de edad es de 42 años, lo que nos habla de una poblacion adulta mayor, considerando los valores minimo y maximo de 27 y 59 años, considerando estos y la desviacion estandar de 8.67, si consideramos la distribucion del promedio hacia su desviacion estandar tendriamos un rango de edad entre 33-50 años de edad, que podemos considera como poblacion adulta.

2. *Sleep Duration*: El promedio de horas de sueño es de 7 hrs, con una desviacion de 0.7 hrs, considerando nuestro valor maximo de 8.5 hrs, estos valores podrian deberse a la edad de nuestra poblacion de edad mayor, por lo que el tiempo de sueño tiende a disminuir con la edad.
3. *Quality Sleep*: La calidad de sueño determinada en escala de 1 a 10, podremos decir que es buena en la mayoria de los casos, teniendo un porcentaje cercano a 7 en la mayor de los casos.
4. *Physical Activity*: Nuestra poblacion se mantiene activa en promedio una hora al dia, como minimo 30 min por dia, lo que indica que toda la poblacion realiza actividad fisica.
5. *Stress Level*: Las personas manejan un estres tolerable a lo largo del dia, con un promedio de 5, y con valores minimos de 3.
6. *Heart Rate*: La poblacion presenta valores normales de frecuencia cardiaca, que se encuentran determinados entre 60 y 100 latidos por minuto, mientras que en nuestos datos tenemos un promedio de 70, y con valores minimos y maximos, dentro del rango normal.
7. *Daily steps*: La cantidad de pasos diarios de las personas, reflejan la actividad fisica que mantienen cotidianamiente, con un minimo de 3000 pasos diarios y un promedio de 6000 pasos.

Analisis de Edades

In [14]:
#cantidad de hombres y mujeres
df['Gender'].value_counts()

Male      189
Female    185
Name: Gender, dtype: int64

In [17]:
#promedio de edades del conjunto
df['Age'].mean()

42.18449197860963

In [18]:
df['Age'].std()
#el conjunto de datos, se inclina a un lado

8.673133465547243

In [19]:
#segmentar en adultos y adultos viejos
df.loc[df['Age']<=40,'Age_group']='Adult'
df.loc[df['Age']>=41,'Age_group']='Older Adult'

In [20]:
df.head()

Gender  Age            Occupation  Sleep Duration  Quality of Sleep  \
Person ID                                                                       
1           Male   27     Software Engineer             6.1                 6   
2           Male   28                Doctor             6.2                 6   
3           Male   28                Doctor             6.2                 6   
4           Male   28  Sales Representative             5.9                 4   
5           Male   28  Sales Representative             5.9                 4   

           Physical Activity Level  Stress Level BMI Category Blood Pressure  \
Person ID                                                                      
1                               42             6   Overweight         126/83   
2                               60             8       Normal         125/80   
3                               60             8       Normal         125/80   
4                               30             8        Obese         140/90   
5                               30             8        Obese         140/90   

           Heart Rate  Daily Steps Sleep Disorder Age_group  
Person ID                                                    
1                  77         4200           None     Adult  
2                  75        10000           None     Adult  
3                  75        10000           None     Adult  
4                  85         3000    Sleep Apnea     Adult  
5                  85         3000    Sleep Apnea     Adult

In [21]:
#conteo de adultos
df['Age_group'].value_counts()
#existen menos adultos, el dataset tiene mas adultos mayores

Older Adult    209
Adult          165
Name: Age_group, dtype: int64

In [22]:
#agrupar por genero
df.groupby('Gender')['Age'].agg(['mean','std'])
#obtener el promedio y desviacion por genero

mean       std
Gender                     
Female  47.405405  8.093407
Male    37.074074  5.662006

Nuestra poblacion tiende a ser mayoritariamente mayor en edad, especialmente en la poblacion de mujeres, por lo que hacer un analisis por medio de la edad para determinar la calidad y tiempo de sueño, podria estar sesgada por la edad y los habitos de las personas mayores

Analisis por ocupacion

In [24]:
#conocer las ocupaciones por genero
df.groupby('Occupation')['Gender'].value_counts()

Occupation            Gender
Accountant            Female    36
                      Male       1
Doctor                Male      69
                      Female     2
Engineer              Female    32
                      Male      31
Lawyer                Male      45
                      Female     2
Manager               Female     1
Nurse                 Female    73
Sales Representative  Male       2
Salesperson           Male      32
Scientist             Female     4
Software Engineer     Male       4
Teacher               Female    35
                      Male       5
Name: Gender, dtype: int64

In [25]:
#tiempo maximo de sueño
print(df['Sleep Duration'].max())

8.5


In [26]:
#calidad de sueño por genero
df.groupby('Gender')[['Sleep Duration','Quality of Sleep']].mean()

Sleep Duration  Quality of Sleep
Gender                                  
Female        7.229730          7.664865
Male          7.036508          6.968254

Tanto hombres como mujeres duermen una cantidad similar de horas, y las recomendadas por la OMS, podemos observar que las mujeres tiene una ligera mejor calidad del sueño

In [28]:
#calidad de sueño por ocupacion
occ_qua = df.groupby('Occupation')[['Sleep Duration','Quality of Sleep']].mean()
occ_qua.sort_values(by='Quality of Sleep')

Sleep Duration  Quality of Sleep
Occupation                                            
Sales Representative        5.900000          4.000000
Scientist                   6.000000          5.000000
Salesperson                 6.403125          6.000000
Software Engineer           6.750000          6.500000
Doctor                      6.970423          6.647887
Teacher                     6.690000          6.975000
Manager                     6.900000          7.000000
Nurse                       7.063014          7.369863
Accountant                  7.113514          7.891892
Lawyer                      7.410638          7.893617
Engineer                    7.987302          8.412698

Debido a la distribucion no equitativa de profesiones en los encuestados, es dificil determinar que profesion presenta un mejor calidad de sueño, pero podemos observa que conforme mas tiempo de sueño se tenga mejor calidad de sueño tendremos. Podemos observar que las profesiones con mejor sueño son Ingenieros, abogados y contadores. Mientras que la peor calidad de sueño la tienen los vendedores

In [30]:
#indice de Masa corporal por ocupacion
occ_bmi = df.groupby('Occupation')['BMI Category'].value_counts()
occ_bmi.sort_values()

Occupation            BMI Category 
Lawyer                Normal Weight     1
Software Engineer     Overweight        1
                      Obese             1
Manager               Overweight        1
Teacher               Obese             1
Doctor                Normal Weight     2
Lawyer                Obese             2
                      Overweight        2
Software Engineer     Normal Weight     2
Sales Representative  Obese             2
Engineer              Overweight        3
                      Normal Weight     4
Doctor                Obese             4
Scientist             Overweight        4
Accountant            Normal Weight     5
                      Overweight        6
Teacher               Normal            6
Nurse                 Normal Weight     7
Accountant            Normal           26
Salesperson           Overweight       32
Teacher               Overweight       33
Lawyer                Normal           42
Engineer              Normal           5

In [8]:
#Nivel de estres por ocupacion
occ_stress = df.groupby('Occupation')['Stress Level'].value_counts()
occ_stress.sort_values()

Occupation            Stress Level
Manager               5                1
Software Engineer     8                1
                      6                1
Engineer              6                1
Software Engineer     5                2
Scientist             8                2
                      6                2
Sales Representative  8                2
Nurse                 6                2
Teacher               5                2
                      6                2
Accountant            3                2
Doctor                5                2
Engineer              7                3
Accountant            6                3
Lawyer                6                3
Nurse                 7                4
                      4                4
Doctor                3                4
Teacher               7                5
Accountant            7                6
Engineer              4                9
                      5               16
Accountant            

Las personas con mayor indice de masa corporal son enfermeras, mientra que la profesion con mayor nivel de estres son los doctores

Analisis por actividad fisica

In [10]:
#Calidad de sueño, paso por ocupacion
occ_steps = df.groupby('Occupation')['Daily Steps','Physical Activity Level','Sleep Duration','Quality of Sleep'].mean()
occ_steps.sort_values(by='Daily Steps')

<ipython-input-10-23df8aa93bed>:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  occ_steps = df.groupby('Occupation')['Daily Steps','Physical Activity Level','Sleep Duration','Quality of Sleep'].mean()


Daily Steps  Physical Activity Level  Sleep Duration  \
Occupation                                                                   
Sales Representative  3000.000000                30.000000        5.900000   
Scientist             5350.000000                41.000000        6.000000   
Manager               5500.000000                55.000000        6.900000   
Software Engineer     5800.000000                48.000000        6.750000   
Teacher               5957.500000                45.625000        6.690000   
Engineer              5980.952381                51.857143        7.987302   
Salesperson           6000.000000                45.000000        6.403125   
Doctor                6808.450704                55.352113        6.970423   
Accountant            6881.081081                58.108108        7.113514   
Lawyer                7661.702128                70.425532        7.410638   
Nurse                 8057.534247                78.589041        7.063014   

                      Quality of Sleep  
Occupation                              
Sales Representative          4.000000  
Scientist                     5.000000  
Manager                       7.000000  
Software Engineer             6.500000  
Teacher                       6.975000  
Engineer                      8.412698  
Salesperson                   6.000000  
Doctor                        6.647887  
Accountant                    7.891892  
Lawyer                        7.893617  
Nurse                         7.369863

Las personas con mas actividad fisica y mayor cantidad de pasos por dia son enfermeras, lo que es esperado debido a la implicacion de las actividades que realizan en esta profesion.

Por su parte, la cantidad de pasos de una persona no esta extrictamente ligada con un tiempo de actividad similar a la cantidad de pasos, por lo que si analizamos la cantidad de paso con respecto al tiempo y calidad podemos, observar que no exites una relacion proporcional entre estos.

In [11]:
occ_steps.sort_values(by='Physical Activity Level')

Daily Steps  Physical Activity Level  Sleep Duration  \
Occupation                                                                   
Sales Representative  3000.000000                30.000000        5.900000   
Scientist             5350.000000                41.000000        6.000000   
Salesperson           6000.000000                45.000000        6.403125   
Teacher               5957.500000                45.625000        6.690000   
Software Engineer     5800.000000                48.000000        6.750000   
Engineer              5980.952381                51.857143        7.987302   
Manager               5500.000000                55.000000        6.900000   
Doctor                6808.450704                55.352113        6.970423   
Accountant            6881.081081                58.108108        7.113514   
Lawyer                7661.702128                70.425532        7.410638   
Nurse                 8057.534247                78.589041        7.063014   

                      Quality of Sleep  
Occupation                              
Sales Representative          4.000000  
Scientist                     5.000000  
Salesperson                   6.000000  
Teacher                       6.975000  
Software Engineer             6.500000  
Engineer                      8.412698  
Manager                       7.000000  
Doctor                        6.647887  
Accountant                    7.891892  
Lawyer                        7.893617  
Nurse                         7.369863

Mientras que la cantidad de actividad fisica realizada en el dia, tiene una mejor relacion con respecto a la calidad de sueño de las personas, por lo que podriamos decir que entre mas cercana sea el tiempo de actividad a una hora por dia, entonces se tendra una mejor calidad de sueño y una mayor duracion de este, en niveles aceptables y recomendados

In [13]:
bmi = df.groupby('BMI Category')[[ 'Heart Rate','Sleep Duration','Quality of Sleep']].mean()
bmi.sort_values(by='Quality of Sleep')

Heart Rate  Sleep Duration  Quality of Sleep
BMI Category                                               
Obese           84.300000        6.960000          6.400000
Overweight      70.945946        6.770270          6.898649
Normal Weight   71.285714        7.333333          7.428571
Normal          68.728205        7.393846          7.661538

Podemos decir que estas variables si estan relacionadas, un indice de masa corporal normal presenta una mejor calidad y tiempo de sueño, y conforme este aumenta, el sueño disminuye en las personas. Por su parte conforme menor sea la frecuencia cardica mejor calidad y tiempo de sueño tendra la persona

In [17]:
bmi_blood = df.groupby('Blood Pressure')[['Sleep Duration','Quality of Sleep']].value_counts()
bmi_blood

Blood Pressure  Sleep Duration  Quality of Sleep
115/75          7.2             8                   20
                7.1             8                   12
115/78          8.0             9                    2
117/76          7.9             8                    1
                6.9             6                    1
                                                    ..
140/95          8.0             9                   11
                8.2             9                    9
                6.2             6                    4
                5.9             6                    1
142/92          7.3             7                    2
Length: 67, dtype: int64

Conforme a los datos, podemos decir que no existe una relacion que demuestre que una presion normal (120/80), mejore la calidad o tiempo de sueño, ni que una presion mayor afecte la calidad de sueño